### ISSUE TITLE
Research new signon method to utilize Google or 3rd party sign on credentials.

### ISSUE THREAD

**Issue by jcockroft64** [jcockroft64/sierrahiking#82]

**Comment by jcockroft64** [comment]
This may be possible.   I see from the  nextend site that you can log in with your Facebook, Google and Twitter accounts and that doesn't require the pro version of nextend.
https://www.wpbeginner.com/plugins/how-to-add-one-click-login-with-google-in-wordpress/

**Comment by russellelliott** [comment]
It definitely is possible. There are several options out there; looking at what they all offer. So far, all the plugins I looked at are available for free in some capacity. The question is what each plugins offer for free. Will provide more info soon.

**Comment by russellelliott** [comment]
# Wordpress Social Login/Register
Here are several Wordpress plugins that offer the ability for users to sign in/register with their Google, Facebook, Twitter, or other accounts. Each of these plugins is able to used for free. Heateor seems like the best choice, as it offers the most authentication methods, and the ability to share site content on many social media platforms. On top of that, it's completely free.

## Heateor Super Socializer Plugin
Website: [https://websensepro.com/blog/how-to-add-facebook-and-google-login-to-wordpress-website/](https://websensepro.com/blog/how-to-add-facebook-and-google-login-to-wordpress-website/)
Website: [https://super-socializer-wordpress.heateor.com/](https://super-socializer-wordpress.heateor.com/)
Video Tutorial: [https://www.youtube.com/watch?v=CB7U5ZrgZjU](https://www.youtube.com/watch?v=CB7U5ZrgZjU)
Plugin Link: [https://wordpress.org/plugins/super-socializer/](https://wordpress.org/plugins/super-socializer/)

This plugin is **100% free**. Each authentication method is set up by entering the App ID and Secret Key from each of the authentication provider's developer consoles. The plugin features step-by-step guides for setting up each authentication method.

Features
- Many Authentication Providers: Facebook, Twitter, Google, Instagram, Yahoo among MANY others (although we probably don't need all of them)
- Each authentication methods adds a button to the Login Page, Registration Page, and Comment Form.
- Social Media Share: allow people to share site content on 100s of social media networks.
- Customize order of authentication buttons on pages
- Sync social media data with Wordpress file

## MiniOrange
Plugin Link: [https://wordpress.org/plugins/miniorange-login-openid/](https://wordpress.org/plugins/miniorange-login-openid/)
Video Tutorial: [https://www.youtube.com/watch?v=-qHgNcHxwsE](https://www.youtube.com/watch?v=-qHgNcHxwsE)
Website: [https://plugins.miniorange.com/social-login-social-sharing](https://plugins.miniorange.com/social-login-social-sharing)

This plugin has a free tier and even has the ability for the authentication methods to be preconfigured by MiniOrange so you don't have to deal with App IDs and Secret Keys. The app also has the ability for user data from multiple 
For each of the authentication methods, there are 2 ways of setting up:
- Configure it yourself
	- Need App ID, Secret Key, and Scope for each of the authentication methods
	- These can be found in that authentication’s developer console. Tutorials will link to that plugin’s tutorial video.
- Make a preconfigured app by making an account with MiniOrange
	- If you choose this, you won’t need to enter a custom App ID or Secret for any of the authentication means you choose.
	- If you decline this, you will get a tutorial for how to set it up yourself
- After saving, you can test to make sure it works.
- Can change positioning of the authentication apps in the plugin menu. Will be reflect in the ordering of the buttons on the login page.

Features
- Free Login options: Facebook, Google, Linkedin, Vkontakte, Twitter, Instagram, Amazon, Salesforce, Windows, Yahoo, etc
- The plugin also offers shortcodes you can use to display the login buttons on other web pages.
- Each authentication methods adds a button to the Login Page, Registration Page, and Comment Form.
- Social Media Share: allow people to share site content on 100s of social media networks.
- Sync user data from multiple social network providers into one account

## NextEnd
Plugin Link: [https://wordpress.org/plugins/nextend-facebook-connect/](https://wordpress.org/plugins/nextend-facebook-connect/)
Website: [https://nextendweb.com/social-login/](https://nextendweb.com/social-login/)
Video Tutorial: [https://www.youtube.com/watch?v=buPTza2-6xc](https://www.youtube.com/watch?v=buPTza2-6xc)
* Free version: Login with Google, Facebook Twitter
    * Each provider has a “get started” button for tutorials of how to get the plugin setup. Need App ID, Secret Key, and Scope for each of the authentication methods
    * People can link and unlink multiple social media profiles from their account.
* Pro version
    * More login options: Apple, Microsoft, Linkedin
    * What would we need? We don’t need all of them
    * Display login buttons in the comments, checkout, etc
	    * Other plugins are able to do this for free
    * Can customize the layout of the login form from several preset options

**Comment by russellelliott** [comment]
Update: I'm experiencing issues trying to set up the preconfigured apps from MiniOrange. The option to make a preconfigured app doesn’t appear for any of the social media profiles. Google, Linkedin, and Amazon are all preconfigured despite me not enabling those yet. I do not have the option to preconfigure those, nor any other social media provider, under a miniOrange app under my account.

Amazon certainly makes sense considering you use AWS and must have some sort of API keys for the site to communicate with it. You might have also mentioned using Google Maps APIs back in Summer. LinkedIn doesn't make sense as you didn't mention using that for the site at all.

I submitted a support ticket to see if they can troubleshoot the issue. If the Amazon and Google APIs currently used for the site interfere with the capability for preconfiguring the other apps via MiniOrange, then doing the preconfigured apps won't be an option.

The other option is to set them all up all the social media authentications manually using each social media's developer platform. MiniOrange does provide tutorials for each so it shouldn't be too hard on your end. Would make sense for it to be under your email, and credit card information if applicable.

**Comment by russellelliott** [comment]
I did find the HTML element that houses the slider that activates the preconfigured app via Inspect element. Maybe I could do something with that?

**Comment by russellelliott** [comment]
Went into inspect element and removed the part that made it invisible. Able to set the slider and the fact it's set up persists on reloads. Should test to make sure it works, though.

**Comment by russellelliott** [comment]
The login with Google, Twitter, and Facebook now appear on the login page.

**Comment by jcockroft64** [comment]
I tested with Google -> logged in a jcockroft96@gmail.com which knows me as "Jason Cockroft".   After i logged in i created a comment.  Funny enough, that comment was instantly posted.  It did not need approval.   I went to check if a new WP user was created.  A new account was not created.

Therefore when you log in with a 3rd party SSO, you can leave comments, but you are not a "subscriber".   I can not change the role of that user to "contributor", "administrator", etc.

**Comment by russellelliott** [comment]
Interesting. Considering your Wordpress account used that same email, that must have logged in as your Jason Cockroft account, seeing as no new account was created and no approval was needed to post the comment. It seems your Google account was successfully linked to your Wordpress account. There's a new column in the users table on the far right with the option to delete that user's social media data, with it saying N/A if no social media linkage was initiated.
![Sample Account Data](https://user-images.githubusercontent.com/77514172/209997035-31a0e234-4878-4e68-adec-3f5e2ffb1827.png)

I made another account on the site with another Google account. It made a new user and sent out an email indicating this. The user has the subscriber role and appears in the list of users. I made a comment in a post and it required approval. Like with your account, it shows the option to delete social media data, implying it is linked to the account.
![New User Data](https://user-images.githubusercontent.com/77514172/209997045-e8d52e55-564d-4412-8ba1-0a352505666e.png)

Also, you can change the default role of a user who registers with a social media account. This setting can be found in the plugin's setting under Social Login -> Registration. The role is currently set to subscriber, but has the ability to be altered by the admins.

MiniOrange appears to work as intended.

**Comment by jcockroft64** [comment]
Very good.   Note the two changes in this image.  (1) change text (2) Center login buttons
<img width="848" alt="Screen Shot 2022-12-29 at 1 48 54 PM" src="https://user-images.githubusercontent.com/53958343/210014549-169243f5-7dc7-45cb-a9b1-7b59093e58f0.png">

**Comment by russellelliott** [comment]
The "Connect With" text can be edited by going into the plugin's settings under Social Login -> Customize Social Media Icons. Text has been changed to "Login with".

Went into the shortcode element editor on the page editor and inserted this custom CSS:
```
/*Put the MiniOrange shortcode and its contents in the center of the div

https://www.w3schools.com/css/css_align.asp*/

.elementor-shortcode {
  margin: auto;
  text-align:center;
}

/*Aligning the text to the center

Overide an !important tag
https://stackoverflow.com/questions/11178673/how-to-override-important
*/

.mo-openid-app-icons>p {
    margin: auto !important;
}

.mo-openid-app-icons>p[style] {margin: auto !important;}
```

Resources:
Align content to center of div
https://www.w3schools.com/css/css_align.asp

Overide an !important tag
https://stackoverflow.com/questions/11178673/how-to-override-important

Centering the buttons was a bit trickier, as I couldn't do it with CSS alone, as the existing `!important` tags kept overriding what I wanted to do with the CSS, even when I added `!important` tags. So, I used Javascript.

Elementor -> Custom Code, then set it to apply to only the login page. Added this custom code.
```
<script>
	/*Change CSS style with Javascript
	https://bobbyhadz.com/blog/javascript-change-style-of-all-elements-with-class
	https://dmitripavlutin.com/foreach-iterate-array-javascript/
	*/
	
	//get the buttons
	var boxes = document.querySelectorAll('.mo_btn-social');
	
	//margin left and right are auto to set them in the middle
	function iterate(item) {
		item.style.marginRight = 'auto';
		item.style.marginLeft = 'auto';
	}
	
	//iterate over each of the buttons
	boxes.forEach(iterate);
</script>
```
Resources:
Change CSS style with Javascript
https://bobbyhadz.com/blog/javascript-change-style-of-all-elements-with-class
https://dmitripavlutin.com/foreach-iterate-array-javascript/

Here Is what the login page looks like now:
<img width="1477" alt="Login Page Layout 12-30-22 1016am" src="https://user-images.githubusercontent.com/77514172/210100729-fc40a353-e8ea-483a-ba24-27b428432006.png">

**Comment by jcockroft64** [comment]
Looks great.

**Comment by jcockroft64** [comment]
![Slide1](https://user-images.githubusercontent.com/53958343/210109170-b1c1bd95-b516-4a17-a811-4a6b61956963.jpeg)

**Comment by jcockroft64** [comment]
I think the current design of the Sign In page is a bit confusing.   Could you separate the Sign In actions from the Registration actions?   I've done a mock up above.

**Comment by russellelliott** [comment]
Custom CSS for the "Sign In" section
```
selector{
    width:80%;
    border-style: solid;
    border-color: #98a790;
    background-color:#fcfef9;
}
```

For the "Register" section. Added 50px of margin on the top to separate the two sections.
```
selector{
    width:80%;
    border-style: solid;
    border-color: #98a790;
    background-color:#fcfef9;
    margin-top:50px;
}
```
"Sign In" and "Register" header colors: #425527; set by selecting the header and going into "Style"

Within each section is a column. I reduced the inner width as to add more padding on the sides to match the mockup.
Here is the custom CSS for that:
```
selector{
    width:80%;
    margin:auto;
}
```
Here is what the login page looks like now:
<img width="1904" alt="Login Page Layout 12-30-22 350pm" src="https://user-images.githubusercontent.com/77514172/210118643-646589cf-9bbf-43d7-a935-cf56055193fb.png">

I used this website to get the colors from the mockup: https://imagecolorpicker.com/en
Border color: https://www.w3schools.com/cssref/pr_border-color.php
`margin: auto` used to center a div within another div
